[elasticsearchexporter] RequireDataStream config

axw opened the pull request:
#### Description

Add configuration for explicitly enabling or disabling `require_data_stream`. If unspecified, defaults to true for "otel" mapping mode and false otherwise.

#### Link to tracking issue

Part of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/36092. Mapping mode may differ across requests, while `require_data_stream` will apply to all requests.

Closes https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37283

#### Testing

Added a unit test.

#### Documentation

Updated README.

Commit:
[elasticsearchexporter] RequireDataStream config

Add configuration for explicitly enabling or disabling
require_data_stream. If unspecified, defaults to true
for "otel" mapping mode and false otherwise.

carsonip commented on exporter/elasticsearchexporter/config.go:
```diff
@@ -62,6 +62,12 @@ type Config struct {
 	// https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest.html
 	Pipeline string `mapstructure:"pipeline"`
 
+	// RequireDataStream controls whether require_data_stream=true is set in
```
Can you explain why anyone wants to explicitly set it to true / false? e.g. < 8.12 doesn't support require_data_stream and will fail every bulk request in OTel mode

axw replied:
> Is this the answer to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37283 or should we look into detecting ES version on startup to avoid the configuration and provide a better UX OOTB?

That was not my motivation, but I suppose it would help. I'm really not keen on detecting ES version and making the behaviour of the exporter dynamic, as long as we can avoid that.

> How will this affect automatic mapping mode work? We set require_data_stream during ES client initialization, which is already a hurdle to implement automatic mapping mode. IIRC later es client versions support per-request require_data_stream, but I'm not sure if it also requires newer ES versions. Will adding such a config increases / decreases the difficulty to do automatic mapping mode?

Why is this a hurdle to automatic mapping mode?

I'm no longer convinced that `require_data_stream` is particularly useful for us. For otel mode, and for Elastic APM data streams, the index templates are set up very early in ES startup. It's still possible that we create data streams using the wrong index templates (e.g. using the base `logs-*` or `metrics-*` templates), so it's not foolproof.

We can set `require_data_stream` per bulk request, but that implies we would not be able to mix mapping modes in a single bulk request. I suppose that might be OK, but it would add complexity. Maybe ES should support `require_data_stream` on a per-bulk operation, if it doesn't already. But again, I don't think it's that useful.

OTOH, by adding a separate config we make the `require_data_stream` behaviour independent of the statically configured mapping mode. So that way we can for example support only writing to data streams while defaulting to "ecs" mapping mode, and still supporting "otel" mode through client metadata.

axw replied:
> I've also been avoiding to detect ES version, but I wonder if we're starting to put too much burden on users. e.g. users using latest EDOT to send to a 8.11 ES (not sure if this setup is supported) will have to dive into the 100+ line long config and set require_data_stream=false at all relevant places. Just something to think about, not a blocker.

Yeah, it's a fair concern. I wouldn't rule it out for the future, for other features. I guess it's inevitable.

> I was assuming that automatic mapping mode means supporting ECS mode with no require_data_stream and OTel mode with require_data_stream at the same time.

I see. If we find there's a need for it in the future, I think we could look at changing the default behaviour of `require_data_stream` to be dependent on the dynamic mapping mode, setting it per-bulk operation.

axw replied:
I just checked, and it turns out bulk operations can already individually specify `require_data_stream`. So what I'm thinking now is:
 - update go-docappender to allow specifying it
 - if you explicitly configure `require_data_stream` then it will enable/disable for all data (that way we can still close https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37283)
 - otherwise it will be dependent on the mapping mode, and will be specified per request

WDYT?

axw replied:
... but apparently support for profiles has been introduced, and some of the docs are not expected to go into data streams at all. I'm increasingly inclined to stop setting this parameter tbh.

rockdaboot replied:
Index templates for profiling are currently set up manually. Ingesting into `profiling-*` DSs and indices without having them set up causes havoc.

One solution would be to set up profiling index and component templates automatically in ES (needs to be discussed).

`require_data_stream` for profiling needs a list of exceptions.

axw replied:
Got it, thanks. I'll close this PR and go back to the drawing board. I'm inclined to stop setting `require_data_stream` everywhere except profiling events in the short term, and work towards having the index templates set up by default.

carsonip reviewed (commented):
code lgtm but I have questions about the motivation though.
- Is this the answer to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/37283 or should we look into detecting ES version on startup to avoid the configuration and provide a better UX OOTB?
- How will this affect automatic mapping mode work? We set require_data_stream during ES client initialization, which is already a hurdle to implement automatic mapping mode. IIRC later es client versions support per-request require_data_stream, but I'm not sure if it also requires newer ES versions. Will adding such a config increases / decreases the difficulty to do automatic mapping mode?

axw committed:
Update exporter/elasticsearchexporter/README.md

Co-authored-by: Carson Ip <carsonip@users.noreply.github.com>

axw committed:
Merge branch 'main' into elasticsearchexporter-requiredatastream

Commit:
Use a single bulk indexer everywhere

Remove the extra bulk indexers added for sending profiles,
and rely on the `require_data_stream` configuration.

axw commented:
@carsonip I dismissed your approval since I'm not longer sure we should even keep setting this parameter.

Commit:
Set per-item require_data_stream

dmathieu commented on exporter/elasticsearchexporter/exporter.go:
```diff
@@ -87,32 +77,11 @@ func (e *elasticsearchExporter) Start(ctx context.Context, host component.Host)
 		return err
 	}
 	e.bulkIndexer = bulkIndexer
-	biEvents, err := newBulkIndexer(e.Logger, client, e.config, true)
```
Having multiple indices here isn't just about require data stream. There's a performance aspect to it as well.

cc @rockdaboot

rockdaboot replied:
Right, that's a learning from the past. Due to certain settings (sorting, less shards), writing events is 8-10x slower than writing stackframes. In the non-OTEL architecture, the different types of documents were delivered in separate messages (think of parallel pipelines). So it made sense to have separate bulk indexers.

Since with the OTEL protocol, all profiling data is included in a single message. So having a single BI possibly doesn't make a difference. Just be aware that the flush operation takes as long as the slowest write.

On a further thought, does it make sense to split bulk indexing by signal? (metrics, logs, profiles having their own BI)

axw replied:
> Since with the OTEL protocol, all profiling data is included in a single message. So having a single BI possibly doesn't make a difference. Just be aware that the flush operation takes as long as the slowest write.

Would it be acceptable if we go back to having a single indexer then, and then do some measurements with this new architecture to see if it's worth having the extra complexity?

> On a further thought, does it make sense to split bulk indexing by signal? (metrics, logs, profiles having their own BI)

Potentially. This is what will happen if you use the `batcher` config (either enabled or explicitly disabled): this disables go-docappender's buffering, and each call to `Consume{Logs,Profiles,Metrics,Traces}` will synchronously encode and end a bulk request to Elasticsearch. Because those are signal-specific, we'll end up getting signal-specific bulk requests.

rockdaboot replied:
Now that we can specify the action type per document, we technically can go back to one BI.

Though we currently have the opportunity to add a latency/duration field to `BulkIndexerResponseStat` and log the average docs/second per BI. Or maybe two fields, as the `Took` value from ES and measured response times on the client side are both interesting. Off course, we can also skip this for now and integrate it into a broader exporter performance investigation or initiative.
